walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,103 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> timesForMeeting = new ArrayList<TimeRange>();
+
+    // Meeting is too long  
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return timesForMeeting;
+    }
+ 
+    // Assume optional attendees can't be included until
+    // slot is found
+    boolean optionalsIncluded = false;
+
+    // Begin search at start of day
+    int beginSlot = TimeRange.START_OF_DAY;
+
+    // Keep track of time slots for optional attendees
+    int beginSlotOptional = beginSlot;
+
+    for (Event event : events) {
+      // Check for conflicts
+      boolean noConflict = checkForConflict(event, request.getAttendees());
+      boolean noOptionalConflict = checkForConflict(event, request.getOptionalAttendees());
+
+      // If optionals can be included do original search method
+      if (optionalsIncluded) {
+        // No conflicts, move on
+        if ( noConflict && noOptionalConflict) {
+          continue;
+        }
+
+        beginSlot = addSlotIfItFits(event, request, beginSlot, timesForMeeting);
+
+      } else {
+        // Check to see if optional attendees can be included
+        if (noOptionalConflict) {
+          if (slotFits(event, request, beginSlotOptional)) {
+            optionalsIncluded = true;
+          }
+        } else {
+          if (slotFits(event, request, beginSlotOptional)) {
+            optionalsIncluded = true;
+            beginSlot = addSlotIfItFits(event, request, beginSlotOptional, timesForMeeting);
+            continue;
+          } else {
+            beginSlotOptional = event.getWhen().end();
+          }
+        }
+
+        // No conflict move on
+        if (noConflict) {
+          continue;
+        }
+
+        beginSlot = addSlotIfItFits(event, request, beginSlot, timesForMeeting);
+      }
+    }
+    
+    // Check for open slot at end of day
+    if (beginSlot < TimeRange.END_OF_DAY) {
+      timesForMeeting.add(TimeRange.fromStartEnd(beginSlot, TimeRange.END_OF_DAY, true));
+    }
+    return timesForMeeting;
+  }
+
+  /* Checks if the set of an event's attendees does not intersect the
+   * set of the request's attendees
+   */
+  public boolean checkForConflict(Event event, Collection<String> attendees) {
+    return Collections.disjoint(event.getAttendees(), attendees);
+  }
+
+  public int addSlotIfItFits(Event event, MeetingRequest request, 
```

Comment:
Should this and checkForConflict be public? If it's not part of your API, make it private. If it is, add a comment to addSlotIfItFits.

Reply:
fixed

Reply:
If it is private it does not need comments?